File: src/NovelRT.Java/core/src/main/java/com/github/novelrt/internal/OwnedHandleObject.java
package com.github.novelrt.internal;

public abstract class OwnedHandleObject extends HandleObject {
  private final HandleContainer handleContainer;

  protected OwnedHandleObject(HandleContainer handlerContainer) {
    this.handleContainer = handlerContainer;
    if (handleContainer.isOwned()) {
      NovelCleaner.CLEANER.register(this, handleContainer.getCleaner());
    }
  }

  protected OwnedHandleObject(long handle, boolean isOwned, HandleDeleter deleter) {
    this(HandleContainer.withDeleter(handle, isOwned, deleter));
  }

  @UsedNatively
  private void invalidate() {
    if (handleContainer.isOwned()) {
      throw new UnsupportedOperationException("Cannot invalidate a OwnedHandleObject that owns its handle.");
    }
    handleContainer.invalidate();
    onInvalidation();
  }

  @Override
  protected @UsedNatively long getHandle() {
    if (!handleContainer.isValid()) {
      throw new IllegalStateException("This HandleObject has been invalidated.");
    }
    return handleContainer.getHandle();
  }

  protected void onInvalidation() {
  }
}

File: src/NovelRT.Java/core/src/main/kotlin/com/github/novelrt/bridge/handle/OwnedHandleObject.kt
package com.github.novelrt.bridge.handle

import com.github.novelrt.bridge.UsedNatively

abstract class OwnedHandleObject(handle: Handle, deleter: (Handle) -> Unit, isOwned: Boolean = true) : HandleObject() {
  private val handleContainer = HandleContainer(handle, deleter, isOwned)

  private fun invalidate() {
    check(handleContainer.isOwned) { "Cannot invalidate a OwnedHandleObject that owns its handle." }
    handleContainer.invalidate()
    onInvalidation()
  }

  @UsedNatively
  override val handle: Long
    get() {
      check(handleContainer.isValid) { "This HandleObject has been invalidated." }
      return handleContainer.handle
    }

  protected open fun onInvalidation() {}

  init {
    if (handleContainer.isOwned) {
      novelCleaner.register(this, handleContainer.cleaner)
    }
  }

  private class HandleContainer(var handle: Handle, private val deleter: (Handle) -> Unit, val isOwned: Boolean) {
    val isValid: Boolean get() = handle != 0L

    fun invalidate() {
      handle = 0
    }

    private fun delete() {
      if (handle != 0L) {
        deleter(handle)
      }
    }

    private fun clean() {
      if (isValid && isOwned) {
        delete()
      }
    }

    val cleaner: () -> Unit = ::clean
  }
}
